### ISSUE TITLE
trying to pickle tests or judge does not work

### ISSUE THREAD

**Issue by russelljjarvis** [scidash/neuronunit#94]
```python
def plot_db(test,judge,vms):
    import matplotlib.pyplot as plt
    plt.clf()
    matplotlib.use('Agg') 
    matplotlib.style.use('ggplot')
    #for t in tests:
    obs = []
    pre = []
    obs.append(judge.observation)
    pre.append(judge.prediction)
    print('observation {0} test {1}'.format(obs,tests[k]))
    print('prediction {0} test {1}'.format(pre,tests[k]))
```

Pickeling is done by ipyparalllel its necissitated because this is the way workers get method arguments in a call to 
```python 
view.map_sync(evaluate, foo, bar)
```

Interestingly its only `suite.judge` that can't be passed between workers using serialization, `suite.test` is fine.

```python
plot_db(get_neab.suite.tests,get_neab.suite.judge,vms)
```

```python
def evaluate(vms):#This method must be pickle-able for ipyparalllel to work.
    '''
    Inputs: An individual gene from the population that has compound parameters, and a tuple iterator that
    is a virtual model object containing an appropriate parameter set, zipped togethor with an appropriate rheobase
    value, that was found in a previous rheobase search.

    outputs: a tuple that is a compound error function that NSGA can act on.

    Assumes rheobase for each individual virtual model object (vms) has already been found
    there should be a check for vms.rheobase, and if not then error.
    Inputs a gene and a virtual model object.
    outputs are error components.
    '''
    from neuronunit.models import backends
    from neuronunit.models.reduced import ReducedModel
    import quantities as pq
    import numpy as np
    import get_neab

    new_file_path = str(get_neab.LEMS_MODEL_PATH)+str(os.getpid())
    model = ReducedModel(new_file_path,name=str(vms.attrs),backend='NEURON')
    model.load_model()
    assert type(vms.rheobase) is not type(None)
    model.update_run_params(vms.attrs)
    tests = get_neab.suite.tests
    for t in tests:
        print(t.observation, t.describe())
    prepare_tests(get_neab.suite.tests,vms)
    model.update_run_params(vms.attrs)
    score = get_neab.suite.judge(model, stop_on_error = False, deep_error = True)
    plot_db(get_neab.suite.tests,get_neab.suite.judge,vms)
    model.run_number += 1
    # Run the model, then:
    error = []
    vms.score = []

    for my_score in score.sort_key.values.tolist()[0]:
        assert type(my_score) is not type(dict)
        error.append(my_score)
    vms.evaluated = True
    error = [ -1.0 * e for e in error ]
    vms.error = error
    return error[0],error[1],error[2],error[3],error[4],error[5],error[6],error[7],

```
```python
fitnesses = toolbox.map(toolbox.evaluate, copy.copy(vmpop))
```
```python
Traceback (most recent call last):
  File "nsga_parallel.py", line 594, in <module>
    fitnesses = toolbox.map(toolbox.evaluate, copy.copy(vmpop))
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py", line 342, in map_sync
    return self.map(f,*sequences,**kwargs)
  File "<decorator-gen-130>", line 2, in map
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py", line 50, in sync_results
    ret = f(self, *args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py", line 615, in map
    return pf.map(*sequences)
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/remotefunction.py", line 285, in map
    return self(*sequences, __ipp_mapping=True)
  File "<decorator-gen-120>", line 2, in __call__
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/remotefunction.py", line 76, in sync_view_results
    return f(self, *args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/remotefunction.py", line 259, in __call__
    ar = view.apply(f, *args)
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py", line 211, in apply
    return self._really_apply(f, args, kwargs)
  File "<decorator-gen-129>", line 2, in _really_apply
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py", line 50, in sync_results
    ret = f(self, *args, **kwargs)
  File "<decorator-gen-128>", line 2, in _really_apply
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py", line 35, in save_ids
    ret = f(self, *args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py", line 557, in _really_apply
    ident=ident)
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/client/client.py", line 1395, in send_apply_request
    item_threshold=self.session.item_threshold,
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/serialize/serialize.py", line 166, in pack_apply_message
    serialize_object(arg, buffer_threshold, item_threshold) for arg in args))
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/serialize/serialize.py", line 166, in <genexpr>
    serialize_object(arg, buffer_threshold, item_threshold) for arg in args))
  File "/opt/conda/lib/python3.5/site-packages/ipyparallel/serialize/serialize.py", line 112, in serialize_object
    buffers.insert(0, pickle.dumps(cobj, PICKLE_PROTOCOL))
  File "/opt/conda/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 706, in dumps
    cp.dump(obj)
  File "/opt/conda/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 146, in dump
    return Pickler.dump(self, obj)
  File "/opt/conda/lib/python3.5/pickle.py", line 408, in dump
    self.save(obj)
  File "/opt/conda/lib/python3.5/pickle.py", line 520, in save
    self.save_reduce(obj=obj, *rv)
  File "/opt/conda/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 586, in save_reduce
    save(args)
  File "/opt/conda/lib/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/conda/lib/python3.5/pickle.py", line 725, in save_tuple
    save(element)
  File "/opt/conda/lib/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/conda/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 264, in save_function
    self.save_function_tuple(obj)
  File "/opt/conda/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 317, in save_function_tuple
    save(f_globals)
  File "/opt/conda/lib/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/conda/lib/python3.5/pickle.py", line 810, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/conda/lib/python3.5/pickle.py", line 836, in _batch_setitems
    save(v)
  File "/opt/conda/lib/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/conda/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 264, in save_function
    self.save_function_tuple(obj)
  File "/opt/conda/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 317, in save_function_tuple
    save(f_globals)
  File "/opt/conda/lib/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/conda/lib/python3.5/pickle.py", line 810, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/conda/lib/python3.5/pickle.py", line 836, in _batch_setitems
    save(v)
  File "/opt/conda/lib/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/conda/lib/python3.5/pickle.py", line 770, in save_list
    self._batch_appends(obj)
  File "/opt/conda/lib/python3.5/pickle.py", line 794, in _batch_appends
    save(x)
  File "/opt/conda/lib/python3.5/pickle.py", line 495, in save
    rv = reduce(self.proto)
  File "/opt/conda/lib/python3.5/site-packages/sciunit-0.1.5.8-py3.5.egg/sciunit/__init__.py", line 54, in __getstate__
    for key in self.unpicklable:
AttributeError: 'RheobaseTest' object has no attribute 'unpicklable'
```

**Comment by rgerkin** [comment]
@russelljjarvis Could you make the minimal recipe that raises this error?  There is a lot going on here so I don't know how to reproduce it on my end.

**Comment by rgerkin** [comment]
@russelljjarvis I've added a check in *sciunit* for this attribute, so that the whole `unpicklable` check will be skipped if that attribute doesn't exist.  But really, it always should exist so I haven't solved the underlying problem (although your code might work now).  That's why I'd like a recipe so I can tackle this problem once and for all.

**Comment by rgerkin** [comment]
@russelljjarvis 
I pushed some new changes to scidash/neuronunit/dev that might help.  They help ensure that all objects can be found again on unpickling.  

However, I realize there is another issue I need a better grasp of, which is what kinds of things in model initialization (with the NEURON backend) should happen once on the host or one worker and just be pushed passively to others (e.g. simple attributes like a resting potential value) and what kinds of things need to be done independently on each worker.  That will determine the structure and content of set_backend, load_model, etc.  The nature of pickling requires that we put the first category of things in __init__ (and functions it calls) and the second category of things in __new__ (and functions it calls).  This is because both __new__ and __init__ are called on model instantiation, but only __new__ is called on unpickling.  And __new__ cannot simply call __init__, for various reasons.  So in order to separate these things, it would help to understand what things you what acts must occur separately for each worker (e.g. maybe `self.ns = nrn.NeuronSimulation(self.tstop, dt=0.0025)`), so I can make sure those things get walled off from the other things that can be simply copied via pickle/unpickle from one worker to the next.  Can you help with that?

**Comment by russelljjarvis** [comment]
This seems like a very complicated issue. I will need to read over it several times.

**Comment by rgerkin** [comment]
@russelljjarvis 
I think I finally have a solution to all of this.  At least I have gotten past all of the pickling errors, and my computer is running hot working on the optimization.  I don't know it is entirely fixed, but if it doesn't finish or results in an error it doesn't seem to be related to simple pickling problems.  

1) Rebuild your stacks from scidash/docker-stacks/master (I largely did not use the `ParallelPyNeuron` version but will eventually need to add your changes in; you might be able to do it just from the `ParallelPyNeuron` version but I haven't tested that).  This rebuild should involve getting the latest changes to sciunit and neuronunit dev branches.  Or you can mount them to `/home/jovyan/mnt/sciunit` and `/home/jovyan/mnt/neuronunit` if you want to work with the (updated) versions on your local machine.   
2) Update the `scidash` branch of `ParallelPyNeuron` where I have update the `reproduce_error/Dockerfile` to include a few things that are needed.  
3) Checkout the `dev` branch of` rgerkin/informatics_poster` and place it inside `ParallelPyNeuron/reproduce_error` next to the Dockerfile.  This will contain the update `nsga_parallel_no_vm.py` file.  
4) Do `docker build /path/to/ParallelPyNEURON/reproduce_error -t error`
5) Do `docker run -it error`
Optionally, add `-v /path/to/sciunit:/home/jovyan/work/sciunit` and `-v /path/to/neuronunit:/home/jovyan/work/neuronunit` to the lines above to use your local copies of sciunit and neuronunit if they are updated.  

This will at least get you past many of the annoying errors that have been plaguing us, but I don't know if it fully works yet because I don't really know what to look for other than the processor doing a lot of work and not getting any errors about attributes not being found.

**Comment by rgerkin** [comment]
@russelljjarvis Just to update, the `nsga_parallel_no_vm.py` file did run to completion, and `models` and `pop` do contain values that seem like the kind of values that would have been returned by a successful run, but I'm not entirely sure how to check this.  What would be a good line to add to the end of the script to check that the optimization in that file ran successfully?

**Comment by rgerkin** [comment]
@russelljjarvis 
Check that this solution (in the updated sciunit and neuronunit) works for you.

**Comment by russelljjarvis** [comment]
I am up to step 2
2) Update the `scidash` branch of `ParallelPyNeuron` where I have update the `reproduce_error/Dockerfile` to include a few things that are needed.

**Comment by rgerkin** [comment]
@russelljjarvis Any progress?

**Comment by rgerkin** [comment]
Simplifications after you verify that the above formula works:
  
- Checkout a new working copy of `scidash/neuronunit`.  Make a fresh branch called `ghissue94`.  
- Checkout your most up-to-date copy of `russelljjarvis/neuronunit@dev`.  
- Move the `russelljjavis/parallelpyneuron/reproduce_error/Dockerfile` to e.g. `scidash/docker-stacks/ghissue94/Dockerfile`, then optionally delete `russelljjavis/parallelypyneuron`.  
- Copy `nsga_parallel_no_vm.py` and its local dependencies (`get_neab.py` and `model_parameters.py` and any others you find you need) to `scidash/neuronunit/optimization@ghissue94`.
- Edit `scidash/docker-stacks/ghissue94/Dockerfile` so that lines 10,11,12,15 reflect `nsga_parallal_no_vm.py` and its dependencies being in `scidash/neuronunit/optimization` rather than in `rgerkin/informatics_posters/neuronunit/optimization`.  
- Rebuild `scidash/docker-stacks/ghissue94/Dockerfile`.  
- Run it and see if it works.  
- If it does, this branch (scidash/neuronunit@ghissue94) can be merged into scidash/neuronunit@dev (create a pull request).
- The continue adding bits and pieces of `russelljjarvis/neuronunit@dev` to `scidash/neuronunit@ghissue94` and see if it still works.  For this you will probably want to mount your local copy of `scidash/neuronunit@ghissue94` into `/home/jovyan/work/neuronunit` in Docker so you can just edit that instead of rebuilding the Docker image every time.  
- When everything you want to copy has been copied, then the original `russelljjarvis/neuronunit@dev` becomes obsolete and we can just make another pull request from `scidash/neuronunit@ghissue94` to integrate all the relevant parts of your code.

**Comment by russelljjarvis** [comment]
So I think implied in this approach, is to use a slightly older version of /models/backends.py the one from your dev branch of informatics_poster.

If I try to use the new backends I get the error.


```
[0:apply]: 
---------------------------------------------------------------------------OSError                                   Traceback (most recent call last)<string> in <module>()
/home/jovyan/work/neuronunit/neuronunit/optimization/nsga_parallel_no_vm.py in p_imports()
     78     from neuronunit.models import backends
     79     from neuronunit.models.reduced import ReducedModel
---> 80     model = ReducedModel(get_neab.LEMS_MODEL_PATH,name='vanilla',backend='NEURON')
     81     model.load_model()
     82     return
/home/jovyan/work/neuronunit/neuronunit/models/reduced.py in __init__(self, LEMS_file_path, name, backend, attrs)
     20         """
     21         mod.LEMSModel.__init__(self,LEMS_file_path,name=name,
---> 22                                           backend=backend,attrs=attrs)
     23         #super(ReducedModel,self).__init__(LEMS_file_path,name=name,
     24         #                                  backend=backend,attrs=attrs)
/home/jovyan/work/neuronunit/neuronunit/models/__init__.py in __init__(self, LEMS_file_path, name, backend, attrs)
     31 
     32         self.orig_lems_file_path = LEMS_file_path
---> 33         self.create_lems_file(name,attrs)
     34         self.run_defaults = pynml.DEFAULTS
     35         self.run_defaults['nogui'] = True
/home/jovyan/work/neuronunit/neuronunit/models/__init__.py in create_lems_file(self, name, attrs)
     82             self.temp_dir = tempfile.gettempdir()
     83         self.lems_file_path  = os.path.join(self.temp_dir, '%s.xml' % name)
---> 84         shutil.copy2(self.orig_lems_file_path, self.lems_file_path)
     85         if attrs:
     86             self.set_lems_attrs(attrs)
/opt/conda/lib/python3.5/shutil.py in copy2(src, dst, follow_symlinks)
    250         dst = os.path.join(dst, os.path.basename(src))
    251     copyfile(src, dst, follow_symlinks=follow_symlinks)
--> 252     copystat(src, dst, follow_symlinks=follow_symlinks)
    253     return dst
    254 
/opt/conda/lib/python3.5/shutil.py in copystat(src, dst, follow_symlinks)
    190             return _nop
    191 
--> 192     st = lookup("stat")(src, follow_symlinks=follow)
    193     mode = stat.S_IMODE(st.st_mode)
    194     lookup("utime")(dst, ns=(st.st_atime_ns, st.st_mtime_ns),
OSError: [Errno 9] Bad file descriptor: '/home/jovyan/work/neuronunit/neuronunit/optimization/NeuroML2/LEMS_2007One.xml'
```

This is coming from line 191 of nsga_no_vm.py from your. branch of informatics_poster@dev, but __only__ when I try to patch it with the newer backends file.

Is it okay to use the older one?
```
        from neuronunit.models import backends
        from neuronunit.models.reduced import ReducedModel

        model = ReducedModel(get_neab.LEMS_MODEL_PATH,name=str('vanilla'),backend='NEURON')
        model.load_model()
        param_dict = {}
```

**Comment by russelljjarvis** [comment]
Since updating this thread last I may have encountered the source of some, but not all of these problems.

I had set the last line of the issue94 dockerfile to be:
```
ENTRYPOINT ipcluster start -n 8 --profile=default & sleep 5 & /bin/bash & ipython stdout_worker.py &
```
the critical part is  `ipcluster start -n 8 --profile=default`

Its default behavior is to configure the clients/engines working directory then and there to be ```--workdir=`pwd````

Where the value of `pwd` is of course the last value of WORKDIR in the Dockerfile.

I have decided this is not explicit enough, so rather than defining an entry point. I have instead created a bash alias that can be launched when you have navigated to your intended directory.

RUN echo "sc='export IPYTHONDIR=`pwd`;ipcluster start -n 8 --profile=default --workdir=`pwd` & sleep 5; python stdout_worker.py &'" >> ~/.bashrc

However, none of this explains why the workers directories cannot be rewritten by editing sys.path inside a `dview.apply_sync` or a `dview.sync_imports()` call. My sense is that just cannot update sys.path on the clients/engines in a persistent sort of way, and the directory you launch ipcluster from is the one the engines are stuck with. Which should be workable anyway.
